### PR TITLE
TokenRules registerRule() function fails on initial rule registration. (fixes #28)

### LINKED ISSUE #28: TokenRules registerRule() function fails on initial rule registration.

### PR DESCRIPTION
**Pull request by pgev**

### CODE REVIEW

File: contracts/TokenRules.sol [hunk 1]
```diff
@@ -64,17 +64,29 @@ contract TokenRules {
...
+     * 'rulesByNameHash' mappings for pointing to the index with 'rules' array.
+     * Simple usage of uint256 in those mappings does not work, because
+     * for non existing rule name and address it defaults to 0 index,
+     * which is obviously wrong. Before accessing 'rules' array by index
+     * one should check 'exists' field of the struct.
+     */
+    struct RuleIndex {
+        uint256 index;
+        bool exists;
+    }
```
Comment (abhayks1) [inline contracts/TokenRules.sol]: Below approach can avoid extra storage:

We can iterate on rules array and check if input ruleNameHash or ruleAddress exist.
The method name can be findRule similar to findGlobalConstraintIndex.
registerRule is not a frequent operation by organization. So the extra iteration can work.

Let me know your opinion on this.

File: test/token_rules/register_rule.js [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+                    web3.utils.keccak256(aRuleName),
+                ),
+            );
+
+            assert.isOk(
+                await tokenRules.rulesByAddress.call(aRuleAddress),
+            );
+        });
+    });
+});
```
Comment (abhayks1) [inline test/token_rules/register_rule.js]: Just to know we will continue write unit tests for other TokenRules methods like addGlobalConstraint, removeGlobalConstraint etc. It will be different PR.

Comment (pgev) [reply test/token_rules/register_rule.js]: Sure.

File: contracts/TokenRules.sol (hunk 1 above)
Comment (pgev) [reply contracts/TokenRules.sol]: Traversing the array is not a good idea and should be avoided if possible. It costs gas and in case of many rules could fail because of gas limit.

**Comment by benjaminbollen** [comment]
fixes #28

File: contracts/TokenRules.sol (hunk 1 above)
Comment (abhayks1) [reply contracts/TokenRules.sol]: ok. I also think we should also follow the same bool exists approach for searching globalConstraints as same logic applies there also. Until there is a strong reason we should do it. Not a work for this PR but something we can take up as a separate issue. Sorry for not commenting that in previous PR reviews.
